includes/class-events-store.php

```diff
@@ -0,0 +1,592 @@
+	 * @param string $action       Name of action used when the event is registered (unhashed)
+	 * @param string $instance     md5 hash of the event's arguments array, which Core uses to index the `cron` option
+	 * @param bool   $flush_cache  Whether or not to flush internal caches after creating/updating the event
+	 * @return bool
+	 */
+	public function mark_job_completed( $timestamp, $action, $instance, $flush_cache = true ) {
+		$job_id = $this->job_exists( $timestamp, $action, $instance, true );
+
+		if ( ! $job_id ) {
+			return false;
+		}
+
+		return $this->mark_job_record_completed( $job_id, $flush_cache );
+	}
+
+	/**
+	 * Set a job post to the "completed" status
+	 *
+	 * @param int $job_id        ID of job's record
+	 * @param bool $flush_cache  Whether or not to flush internal caches after creating/updating the event
+	 * @return bool
+	 */
+	public function mark_job_record_completed( $job_id, $flush_cache = true ) {
+		global $wpdb;
+
+		$updates = array(
+			'status'   => self::STATUS_COMPLETED,
+			'instance' => wp_rand( 1000000, 999999999 ), // Breaks unique constraint, and can be recreated from entry's remaining data
+		);
+
+		$success = $wpdb->update( $this->get_table_name(), $updates, array( 'ID' => $job_id, ) );
+
+		// Delete internal cache
+		// Should only be skipped during bulk operations
+		if ( $flush_cache ) {
+			$this->flush_internal_caches();
+		}
+
+		return (bool) $success;
+	}
+
+	/**
+	 * Compare two arrays and return collapsed representation of their differences
+	 *
+	 * @param array $new New cron array
+	 * @param array $old Old cron array
+	 *
+	 * @return array
+	 */
+	private function find_cron_array_differences( $new, $old ) {
+		$differences = array();
+
+		$new = collapse_events_array( $new );
+
+		foreach ( $new as $event ) {
+			$event = (object) $event;
+
+			if ( ! isset( $old[ $event->timestamp ][ $event->action ][ $event->instance ] ) ) {
+				$differences[] = array(
+					'timestamp' => $event->timestamp,
+					'action'    => $event->action,
+					'instance'  => $event->instance,
+					'args'      => $event->args,
+				);
+			}
+		}
+
+		return $differences;
+	}
+
+	/**
+	 * Delete the cached representation of the cron option
+	 */
+	public function flush_internal_caches() {
+		return wp_cache_delete( self::CACHE_KEY );
+	}
+
+	/**
+	 * Prevent event store from creating new entries
+	 *
+	 * Should be used sparingly, and followed by a call to resume_event_creation(), during bulk operations
+	 */
+	public function suspend_event_creation() {
+		$this->job_creation_suspended = true;
+	}
+
+	/**
+	 * Stop discarding events, once again storing them in the table
+	 */
+	public function resume_event_creation() {
+		$this->job_creation_suspended = false;
+	}
+
+	/**
+	 * Remove entries for non-recurring events that have been run
+	 */
+	public function purge_completed_events( $count_first = true ) {
+		global $wpdb;
+
+		// Skip count if already performed
+		if ( $count_first ) {
```

Comment:
When it's called from CLI, the count query is already run, so I wanted to avoid running it again or blocking the delete call. Now that I've described it, it should be `$force` instead.